sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -760,12 +757,43 @@ object CollapseProject extends Rule[LogicalPlan] {
       s.copy(child = p2.copy(projectList = buildCleanedProjectList(l1, p2.projectList)))
   }
 
+  private def collapseProjects(plan: LogicalPlan): LogicalPlan = plan match {
+    case p1 @ Project(_, p2: Project) =>
+      val maxCommonExprs = SQLConf.get.maxCommonExprsInCollapseProject
+
+      if (haveCommonNonDeterministicOutput(p1.projectList, p2.projectList) ||
+          getLargestNumOfCommonOutput(p1.projectList, p2.projectList) > maxCommonExprs) {
```

Comment:
How about following the `ScanOperation` condition like this?
```
    case p1 @ Project(_, p2: Project) =>
      val maxCommonExprs = SQLConf.get.maxCommonExprsInCollapseProject

      if (haveCommonNonDeterministicOutput(p1.projectList, p2.projectList) ||
          moreThanMaxAllowedCommonOutput(p1.projectList, p2.projectList)) {
```

Reply:
Good. Updated.